File: task-manager-ios/task-manager-ios/Presentation/Core/TasksListState.swift
//
//  TasksListState.swift
//  task-manager-ios
//
//  Created by Jeofferson Dela Peña on 3/9/24.
//

import SwiftUI

struct TasksListState {
    enum SheetType {
        case add
        case modify(Task)
    }
    
    enum AlertType {
        case deleteConfirmation(Task)
        case error(TaskManagerError)
    }
    
    var tasks: [Task] = []
    var completedTasks: [Task] {
        tasks.filter { task in
            task.isCompleted
        }
    }
    
    var isLoading = false
    
    var activeSheet: Binding<SheetType?> = .constant(nil)
    var activeAlert: Binding<AlertType?> = .constant(nil)
}

File: task-manager-android/app/src/main/java/com/jeoffersondelapena/task_manager_android/presentation/core/TasksListState.kt
package com.jeoffersondelapena.task_manager_android.presentation.core

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import com.jeoffersondelapena.task_manager_android.domain.model.Task
import com.jeoffersondelapena.task_manager_android.domain.util.helper.TaskManagerException

object TasksListState {
    sealed class ModalBottomSheetType {
        data object Add : ModalBottomSheetType()

        data class Modify(val task: Task) : ModalBottomSheetType()
    }

    sealed class DialogType {
        data class DeleteConfirmation(val task: Task) : DialogType()

        data class Error(val exception: TaskManagerException) : DialogType()
    }

    var tasks: MutableState<List<Task>> = mutableStateOf(listOf())
    val completedTasks: List<Task>
        get() {
            return tasks.value.filter { task ->
                task.isCompleted
            }
        }

    var isLoading = false

    var activeModalBottomSheet: MutableState<ModalBottomSheetType?> = mutableStateOf(null)
    var activeDialog: MutableState<DialogType?> = mutableStateOf(null)
}